### ISSUE TITLE
[analyzer] Needs to determine the public library to use to reference elements defined in private libraries

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/sdk#55480]
The problem we currently face is that when a tool needs to add an import for an element defined in a private library we have no way to add an import for anything other than the defining (private) library, even if that element is also exported by a public library.

The need is to be able to take an arbitrary element from an internal library and efficiently determine which public libraries, if any, make that element available. Individual tools could then decide how to handle the cases where there are zero, one, or many such libraries.

We minimally need this for code completion and quick fixes / quick assists / code actions. If we're going to support it for quick fixes we should implement this in such a way that plugins can also use this API.

The dartdoc implementation currently has the notion of a 'canonical' library. We'd like to unify these two pieces of implementation into a common set of APIs, though the concept of choosing one of the potentially many public libraries to be the canonical library might only be useful to dartdoc.

The question, then, is what design we ought to use for this piece of support. The purpose of this issue is to discuss whether there are other requirements and what that design might look like.

**Comment by bwilkerson** [comment]
@srawlins @kallentu @DanTup @scheglov

**Comment by srawlins** [comment]
> The problem we currently face is that when a tool needs to add an import for an element defined in a private library we have no way to add an import for anything other than the defining (private) library, even if that element is also exported by a public library.

Hmm, I thought we had that implementation in analyzer already? Well, two:

* [ImportLibrary._importLibraryForElement](https://github.com/dart-lang/sdk/blob/2c0a33400e21ba77bc0724c15786704661c12891/pkg/analysis_server/lib/src/services/correction/dart/import_library.dart#L241)
* [NotImportedContributor.computeSuggestions](https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/services/completion/dart/not_imported_contributor.dart#L34)
* Some shared code at https://github.com/dart-lang/sdk/blob/main/pkg/analyzer/lib/src/dart/analysis/file_state_filter.dart

**Comment by scheglov** [comment]
I don't think that `exported by` relationship can work. The index is per Dart file, it can have only information that is available in this file and its dependencies. It by definition cannot have information that a `class A` declared in this library is re-exported by another library. We cannot go back and patch the index of a library.

We already have this information about exports: `LibraryElement2.exportNamespace`. And `NotImportedCompletionPass` already uses it, by determining which libraries are available to the completion target, and looks into their `exportNamespace`. From this, there is one step to add, if necessary, a caching layer.
